File: chapter03/examples/QtCone.py
#!/usr/bin/env python
#
# This example demonstrates how to embed a VTK render window inside a
# Qt application using PySide6. A cone is rendered in a
# QVTKRenderWindowInteractor widget placed as the central widget of a
# QMainWindow.
#

import sys

# Use QOpenGLWidget as the base class for the VTK widget.  This must
# be set before importing QVTKRenderWindowInteractor.
import vtkmodules.qt
vtkmodules.qt.QVTKRWIBase = "QOpenGLWidget"

from PySide6.QtWidgets import QApplication, QMainWindow

from vtkmodules.qt.QVTKRenderWindowInteractor import QVTKRenderWindowInteractor
from vtkmodules.vtkFiltersSources import vtkConeSource
from vtkmodules.vtkRenderingCore import (
    vtkActor,
    vtkPolyDataMapper,
    vtkRenderer,
)

# Ensure an OpenGL rendering backend and interactor style are loaded
import vtkmodules.vtkRenderingOpenGL2  # noqa: F401
import vtkmodules.vtkInteractionStyle  # noqa: F401

app = QApplication(sys.argv)
window = QMainWindow()
window.setWindowTitle("VTK Qt Cone Example")
window.resize(800, 600)

# Create the VTK widget and place it in the main window.
vtk_widget = QVTKRenderWindowInteractor(window)
window.setCentralWidget(vtk_widget)

# Build a VTK pipeline.
cone = vtkConeSource()
cone.SetHeight(3.0)
cone.SetRadius(1.0)
cone.SetResolution(30)

mapper = vtkPolyDataMapper()
mapper.SetInputConnection(cone.GetOutputPort())

actor = vtkActor()
actor.SetMapper(mapper)

renderer = vtkRenderer()
renderer.AddActor(actor)
renderer.SetBackground(0.1, 0.2, 0.4)

vtk_widget.GetRenderWindow().AddRenderer(renderer)

window.show()
vtk_widget.Initialize()
vtk_widget.Start()

sys.exit(app.exec())

File: chapter03/examples/QtConeCxx/QtCone.cxx
/*=========================================================================

  Program:   Visualization Toolkit
  Module:    QtCone.cxx

  Copyright (c) Ken Martin, Will Schroeder, Bill Lorensen
  All rights reserved.
  See Copyright.txt or http://www.kitware.com/Copyright.htm for details.

     This software is distributed WITHOUT ANY WARRANTY; without even
     the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
     PURPOSE.  See the above copyright notice for more information.

=========================================================================*/
//
// This example demonstrates how to embed a VTK render window inside a
// Qt application. A cone is rendered in a QVTKOpenGLNativeWidget placed
// as the central widget of a QMainWindow.
//

#include <QApplication>
#include <QMainWindow>
#include <QSurfaceFormat>
#include <QVTKOpenGLNativeWidget.h>

#include <vtkActor.h>
#include <vtkConeSource.h>
#include <vtkGenericOpenGLRenderWindow.h>
#include <vtkNew.h>
#include <vtkPolyDataMapper.h>
#include <vtkRenderer.h>

int main(int argc, char* argv[])
{
  // Set the default surface format BEFORE constructing QApplication.
  QSurfaceFormat::setDefaultFormat(QVTKOpenGLNativeWidget::defaultFormat());
  QApplication app(argc, argv);

  // Create a QMainWindow with a VTK render widget.
  QMainWindow window;
  window.setWindowTitle("VTK Qt Cone Example");
  window.resize(800, 600);

  QVTKOpenGLNativeWidget* vtkWidget = new QVTKOpenGLNativeWidget();
  window.setCentralWidget(vtkWidget);

  // Prevent the macOS trackpad from generating spurious button-press
  // events via touch input.  Normal mouse/trackpad clicks still work.
  vtkWidget->setAttribute(Qt::WA_AcceptTouchEvents, false);

  // Create a vtkGenericOpenGLRenderWindow and assign it to the widget.
  vtkNew<vtkGenericOpenGLRenderWindow> renderWindow;
  vtkWidget->setRenderWindow(renderWindow);

  // Build a VTK pipeline.
  vtkNew<vtkConeSource> cone;
  cone->SetHeight(3.0);
  cone->SetRadius(1.0);
  cone->SetResolution(30);

  vtkNew<vtkPolyDataMapper> mapper;
  mapper->SetInputConnection(cone->GetOutputPort());

  vtkNew<vtkActor> actor;
  actor->SetMapper(mapper);

  vtkNew<vtkRenderer> renderer;
  renderer->AddActor(actor);
  renderer->SetBackground(0.1, 0.2, 0.4);

  renderWindow->AddRenderer(renderer);

  window.show();

  // Qt's event loop replaces vtkRenderWindowInteractor::Start().
  return app.exec();
}
